crates/solver/src/settlement_access_list.rs

```diff
@@ -74,22 +100,81 @@ impl TenderlyApi {
             .json()
             .await
     }
+
+    async fn block_number(&self, network_id: String) -> reqwest::Result<BlockNumber> {
+        self.client
+            .get(format!(
+                "https://api.tenderly.co/api/v1/network/{}/block-number",
+                network_id
+            ))
+            .send()
+            .await?
+            .error_for_status()?
+            .json()
+            .await
+    }
+}
+
+#[async_trait::async_trait]
+impl AccessListEstimating for TenderlyApi {
+    async fn estimate_access_lists(
+        &self,
+        txs: &[TransactionBuilder<DynTransport>],
+    ) -> Vec<Result<AccessList>> {
+        futures::future::join_all(txs.iter().map(|tx| async {
+            let input = tx
+                .data
+                .clone()
+                .ok_or(anyhow!("transaction data does not exist"))?;
+            let from = tx
+                .from
+                .clone()
+                .ok_or(anyhow!("transaction from does not exist"))?
+                .address();
+            let to = tx.to.ok_or(anyhow!("transaction to does not exist"))?;
+            let block_number = self.block_number(self.network_id.clone()).await?;
+
+            let tenderly_request = TenderlyRequest {
+                network_id: self.network_id.clone(),
+                block_number: block_number.block_number,
+                from,
+                input,
+                to,
+                generate_access_list: true,
+            };
+
+            match self.access_list(tenderly_request).await {
+                Ok(response) if response.generated_access_list.is_empty() => {
+                    Err(anyhow!("empty access list"))
+                }
+                Ok(response) => Ok(response
+                    .generated_access_list
+                    .into_iter()
+                    .map(Into::into)
+                    .collect()),
+                Err(err) => Err(err.into()),
+            }
```

Comment:
ubernit: A slightly more succinct way to express this with less "right-ward drift" would be to use `?` and `ensure`:
```suggestion
            let response = self.access_list(tenderly_request).await?;
            ensure!(!response.generated_access_list.is_empty(), "empty access list");
            Ok(response
                .generated_access_list
                .into_iter()
                .map(Into::into)
                .collect())
```

Reply:
also nice :)